dds/DCPS/DataReaderImpl_T.h

```diff
@@ -311,10 +312,10 @@ namespace OpenDDS {
 
     const Observer_rch observer = get_observer(Observer::e_SAMPLE_TAKEN);
 
-    const typename InstanceMap::iterator the_end = instance_map_.end();
-    for (typename InstanceMap::iterator it = instance_map_.begin(); it != the_end; ++it) {
-      DDS::InstanceHandle_t handle = it->second;
-      OpenDDS::DCPS::SubscriptionInstance_rch ptr = get_handle_instance(handle);
+    for (SubscriptionInstanceMapType::iterator iter = instances_.begin();
```

Comment:
@mitza-oci what's the difference between instances_ and instance_map_? It looks like we're avoiding get_handle_instance() altogether this way (and the weakptr lock / lookup), but I can't tell if that's a good thing or a bad thing. I think you touched the code for these handles most recently, right?

Reply:
`instances_` is `InstanceHandle_t` -> `SubscriptionInstance_rch` and comes from the non-templated base class
`instance_map_` is `MessageType` -> `InstanceHandle_t` and must be in the template since it uses `MessageType`

Reply:
One of the problems I saw was that traversing the instance_map_ and then calling get_handle_instance() led to inconsistent ordering when  taking collections of samples. For instance the subscriberimpl::get_datareaders() function iterates over the instances_ map, and may evaluate instances with handles 7 and 9, but then the datareaderimpl::take() starts out with instance handle 8.